core/src/test/scala/akka/persistence/jdbc/query/JournalDaoStreamMessagesMemoryTest.scala

```diff
@@ -0,0 +1,157 @@
+  val journalTableCfg = journalConfig.journalTableConfiguration
+
+  import profile.api._
+
+  implicit val askTimeout = 50.millis
+
+  def generateId: Int = 0
+
+  val memoryMBean: MemoryMXBean = ManagementFactory.getMemoryMXBean
+
+  behavior.of("Replaying Persistence Actor")
+
+  it should "stream events" in {
+    withActorSystem { implicit system: ActorSystem =>
+      withDatabase { db =>
+        implicit val mat: ActorMaterializer = ActorMaterializer()
+        implicit val ec: ExecutionContextExecutor = system.dispatcher
+
+        val persistenceId = UUID.randomUUID().toString
+        val dao = new ByteArrayJournalDao(db, profile, journalConfig, SerializationExtension(system))
+
+        val payloadSize = 5000 // 5000 bytes
+        val eventsPerBatch = 1000
+
+        val maxMem = 64 * MB
+
+        val numberOfInsertBatches = {
+          // calculate the number of batches using a factor to make sure we go a little bit over the limit
+          (maxMem / (payloadSize * eventsPerBatch) * 1.2).round.toInt
+        }
+        val totalMessages = numberOfInsertBatches * eventsPerBatch
+        val totalMessagePayload = totalMessages * payloadSize
+        log.info(
+          s"batches: $numberOfInsertBatches (with $eventsPerBatch events), total messages: $totalMessages, total msgs size: $totalMessagePayload")
+
+        // payload can be the same when inserting to avoid unnecessary memory usage
+        val payload = Array.fill(payloadSize)('a'.toByte)
+
+        val lastInsert =
+          Source
+            .fromIterator(() => (1 to numberOfInsertBatches).toIterator)
+            .mapAsync(1) { i =>
+              val end = i * eventsPerBatch
+              val start = end - (eventsPerBatch - 1)
+              log.info(s"batch $i - events from $start to $end")
+              val atomicWrites =
+                (start to end).map { j =>
+                  AtomicWrite(immutable.Seq(PersistentRepr(payload, j, persistenceId)))
+                }.toSeq
+
+              dao.asyncWriteMessages(atomicWrites).map(_ => i)
+            }
+            .runWith(Sink.last)
+
+        // wait until we write all messages
+        // being very generous, 1 second per message
+        lastInsert.futureValue(Timeout(totalMessages.seconds))
+
+        log.info("Events written, starting replay")
+
+        Thread.sleep(1000)
+        System.gc()
+        Thread.sleep(1000)
+        val usedBefore = memoryMBean.getHeapMemoryUsage.getUsed
+
+        val messagesSrc =
+          dao.messagesWithBatch(persistenceId, 0, totalMessages, batchSize = 100, None)
+        val probe =
+          messagesSrc
+            .map {
+              case Success(repr) =>
+                if (repr.sequenceNr % 100 == 0)
+                  log.info(s"fetched: ${repr.persistenceId} - ${repr.sequenceNr}/${totalMessages}")
+              case Failure(exception) => println(exception)
+            }
+            .runWith(TestSink.probe)
+
+        probe.request(10)
```

Comment:
I changed the test to no rely on small heap. Requesting 10 shouldn't fill up memory with all/many records. This works as expected before and after the fix.